oak_functions/loader/src/server.rs

```diff
@@ -461,19 +452,42 @@ impl wasmi::Externals for WasmState {
                     .read_extension_args(request_ptr, request_len)
                     .map_err(|err| {
                         self.log_error(&format!(
-                            "Handle {:?}: Unable to read input from guest memory: {:?}",
-                            extension.get_handle(),
+                            "Unable to read input from guest memory: {:?}",
                             err
                         ));
                         OakStatus::ErrInvalidArgs
                     });
 
+                let extension = &mut self
+                    .extensions_indices
+                    .get_mut(&index)
+                    .unwrap_or_else(|| panic!("Unimplemented function at {}", index));
+
                 let result = match request {
-                    Ok(request) => extension.invoke(self, args, request)?,
+                    Ok(request) => {
+                        let response = extension.invoke(request);
+                        match response {
+                            Ok(response) => {
+                                if !response.is_empty() {
+                                    // Careful: We assume every ABI call which returns a non-emtpy
+                                    // response provides these arguments
+                                    // (which is true).
+                                    let response_ptr_ptr: AbiPointer = args.nth_checked(2)?;
+                                    let response_len_ptr: AbiPointer = args.nth_checked(3)?;
+                                    self.write_extension_result(
+                                        response,
+                                        response_ptr_ptr,
+                                        response_len_ptr,
+                                    )
+                                } else {
+                                    Ok(())
```

Comment:
It looks like here it is not writing anything to the Wasm memory, which seems wrong? It should at least write a zero in the response_len_ptr, right? Otherwise how would the caller know? It should just write an empty vector -- and if there is special logic for empty vectors, that should be handled inside `write_extension_result`, not here.

Reply:
This code is actually called through the soon-to-be-deprecated ABI functions `get_storage_item`, `tf_infer`, and `report_metric`. 

The latter does not write anything to memory (has no result), hence in does not currently have `args.nth_checked(2-3)`.

We write an empty vector for the [testing extension here](https://github.com/project-oak/oak/blob/9c639b300e7e2a6d2afe63c16bc89de6e2b9ac25/oak_functions/loader/src/testing.rs#L44), but the testing extension is already called via `invoke`, i.e., [`invoke_extension_with_handle`](https://github.com/project-oak/oak/blob/9c639b300e7e2a6d2afe63c16bc89de6e2b9ac25/oak_functions/loader/src/server.rs#L388).

Reply:
Uhmm ok I'm still a bit confused by what code is temporary and what code is here to stay, but I trust it will become clearer down the line.